### PR TITLE
Implement IDownstreamApi overloads that take JsonTypeInfo<T> as a parameter to enable source generated Json deserialization for NativeAOT

### PR DESCRIPTION
**Pull request by dualtagh**
Addresses #2930

- Update Microsoft.Identity.Abstractions to 7.0.0 
- Implementation for new DownstreamApi interfaces in Microsoft.Identity.Abstractions for source gen Json serialization. Implementaitons are generated or copied as the existing with just the addition of jsonTypeInfo params

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -111,6 +111,7 @@
     <SystemFormatsAsn1Version>$(NetNineRuntimeVersion)</SystemFormatsAsn1Version>
     <SystemTextJsonVersion>$(NetNineRuntimeVersion)</SystemTextJsonVersion>
     <MicrosoftExtensionsDependencyInjectionVersion>$(NetNineRuntimeVersion)</MicrosoftExtensionsDependencyInjectionVersion>
+    <MicrosoftIdentityAbstractionsVersion>7.0.0</MicrosoftIdentityAbstractionsVersion>
```
Comment (jmprieur) [inline Directory.Build.props:114]: What happens if we reference IdentityAbstractions 7.0.0 in other frameworks? this should work too given the methods in Abstractions are defined only above 8.0.0?

I think we should systematically reference IdentityAbstractions 7.0.0

File: tests/Microsoft.Identity.Web.Test/Microsoft.Identity.Web.Test.csproj [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
   <ItemGroup>
     <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
-    <PackageReference Include="Microsoft.Identity.Abstractions" Version="$(MicrosoftIdentityAbstractionsVersion)" />
```
Comment (jmprieur) [inline tests/Microsoft.Identity.Web.Test/Microsoft.Identity.Web.Test.csproj:23]: I would keep that

File: Directory.Build.props (hunk 1 above)
Comment (jmprieur) [reply Directory.Build.props:114]: Actually I checked. Just update line 91.
Let's just factorize in the common properties. No special case for net8.0 and net.9.0

**Review by jmprieur — APPROVED** [review]
I approved but requested that you don't special case the reference for net8.0 and net9.0 in the Directory.Build.props, but just use Abstractions 7.0.0 for all target framework (change line 91).

Otherwise this will break our unified / id4sbuild
cc: @jennyf19

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs [hunk 3]
```diff
@@ -219,7 +302,7 @@ public Task<HttpResponseMessage> CallApiForAppAsync(
             return clonedOptions;
         }
 
-        internal static HttpContent? SerializeInput<TInput>(TInput input, DownstreamApiOptions effectiveOptions)
+        internal static HttpContent? SerializeInput<TInput>(TInput input, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TInput>? inputJsonTypeInfo = null)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:305]: Adding an optional parameter is a binary breaking change. I think this should be an additional overload and add it to NET8 only section.

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs [hunk 4]
```diff
@@ -234,17 +317,23 @@ public Task<HttpResponseMessage> CallApiForAppAsync(
...
+                        inputJsonTypeInfo == null ? JsonSerializer.Serialize(input) : JsonSerializer.Serialize(input, inputJsonTypeInfo),
+                        Encoding.UTF8,
+                        "application/json"),
                 };
             }
             return httpContent;
         }
 
-        internal static async Task<TOutput?> DeserializeOutput<TOutput>(HttpResponseMessage response, DownstreamApiOptions effectiveOptions)
+        internal static async Task<TOutput?> DeserializeOutput<TOutput>(HttpResponseMessage response, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TOutput>? outputJsonTypeInfo = null)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:336]: Same comment as above.

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs
Comment (pmaytak) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:1]: I assume you won't be using `CallApiAsync`, so not adding overloads for them for now?

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs (hunk 3 above)
Comment (dualtagh) [reply src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs]: @pmaytak Thanks, I've updated it to have an overload and keep the original signature. I've put the implementation into a different method though - is this enough re binary breaking change for an internal method? Or does the original implementation need to stay the same?

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs [hunk 5]
```diff
@@ -166,8 +167,101 @@ public Task<HttpResponseMessage> CallApiForAppAsync(
...
+            where TOutput : class
+        {
+            DownstreamApiOptions effectiveOptions = MergeOptions(serviceName, downstreamApiOptionsOverride);
+            HttpResponseMessage response = await CallApiInternalAsync(serviceName, effectiveOptions, true,
+                                                                          null, null, cancellationToken).ConfigureAwait(false);
+
+            return await DeserializeOutput<TOutput>(response, effectiveOptions, outputJsonTypeInfo).ConfigureAwait(false);
+        }
+
+        internal static HttpContent? SerializeInput<TInput>(TInput input, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TInput>? inputJsonTypeInfo)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:253]:
```suggestion
        internal static HttpContent? SerializeInput<TInput>(TInput input, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TInput> inputJsonTypeInfo)
```

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs [hunk 6]
```diff
@@ -166,8 +167,101 @@ public Task<HttpResponseMessage> CallApiForAppAsync(
...
+        {
+            return SerializeInputImpl(input, effectiveOptions, inputJsonTypeInfo);
+        }
+
+        internal static async Task<TOutput?> DeserializeOutput<TOutput>(HttpResponseMessage response, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TOutput>? outputJsonTypeInfo = null)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:258]: To match SerializeInput signature:
```suggestion
        internal static async Task<TOutput?> DeserializeOutput<TOutput>(HttpResponseMessage response, DownstreamApiOptions effectiveOptions, JsonTypeInfo<TOutput> outputJsonTypeInfo)
```

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs (hunk 3 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs]: I think this is good.

**Review by pmaytak — APPROVED** [review]

**Review by jennyf19 — APPROVED** [review]
Great to see the additional test coverage @dualtagh 
:shipit:

**Pull request merged**